src/stockwatch/entities.py

```diff
@@ -24,6 +25,74 @@ class ShareTransactionKind(Enum):
     DIVIDEND = auto()
 
 
+@dataclass(frozen=False, order=True)
+class CurrencyExchange:  # pylint: disable=too-many-instance-attributes
+    """For representing an exchange at a certain date.
+
+    Attributes
+    ----------
+    exchange_datetime     : the date for which the exchange is registered
+    exchange_rate         : the rate of exchange applied in the transaction
+    value_from            : the amount before the exchange, in curr_from (usually negative)
+    curr_from             : the currency shorthand before the exchange, e.g. USD
+    exchange_rate_exact   : the rate of exchange that is exactly value_to / -value_from (no init)
+    value_to              : the amount after the exchange, in curr_to (no init)
+    value_trans           : the amount that was traced back to transactions (no init)
+    curr_to               : the currency shorthand after the exchange; always EUR (no init)
+    """
+
+    exchange_datetime: datetime
+    exchange_rate: float
+    value_from: float
+    curr_from: str
+    exchange_rate_exact: float = field(init=False)
+    value_to: float = field(init=False)
+    value_trans: float = field(init=False)
+    curr_to: str = field(init=False)
+
+    def __post_init__(self) -> None:
+        assert self.exchange_rate > 0.0
+        assert abs(self.value_from) > 0.0
+        self.value_to = round(-self.value_from / self.exchange_rate, 2)
+        self.curr_to = "EUR"
+        self.exchange_rate_exact = -self.value_from / self.value_to
+        self.value_trans = 0.0
+
+    @property
+    def exchange_date(self) -> date:
+        "Exchange date"
+        return self.exchange_datetime.date()
+
+    @property
+    def value_trans_remaining(self) -> float:
+        "The part of -value_from that has not yet been traced to a transaction"
+        return -self.value_from - self.value_trans
+
+    def has_been_traced_fully(self) -> bool:
+        "Return True if the amount traced to transactions equals the value_to"
+        # we'll allow for a little margin
+        return abs(self.value_trans_remaining) < ZERO_MARGIN
+
+    def can_take_exchange_value(self, the_value: float) -> bool:
+        "Return True if the_value fits in value_trans_remaining"
+        if self.has_been_traced_fully():
+            return False
+        # the_value must have the same sign as value_trans_remaining
+        if the_value * self.value_trans_remaining < 0.0:
+            return False
+        if the_value < 0.0:
+            # we'll allow for a (positive) little margin
+            return self.value_trans_remaining - the_value < ZERO_MARGIN
+        # we'll allow for a (negative) little margin
+        return self.value_trans_remaining - the_value > -ZERO_MARGIN
```

Comment:
No that is not equivalent. Eg: value_trans_remaining -10, the_value -5; the current test passes, your proposition fails

Reply:
Haha, okay yeah I had the condition the wrong way around. So:
```
return abs(self.value_trans_remaining - the_value) > ZERO_MARGIN
```

The only real difference is then whether or not we explicitly want to always 'go over' 0. So in the proposed case value_trans_remaining := 1.0 and the_value 0.99 would return `False`, while in the case above it would return `True`. Is there a reason we would want this margin such that we always use more then defined, instead of a margin going both ways?